client/js/serverstate.js

```diff
@@ -108,7 +108,7 @@ export function batchEnd() {
 function receiveDelta(delta) {
   // the order of widget changes is not necessarily correct and in order to avoid cyclic children, this first moves affected widgets to the top level
   for(const widgetID in delta.s) {
-    if(delta.s[widgetID] && delta.s[widgetID].parent !== undefined && widgets.has(widgetID)) {
+    if(delta.s[widgetID] && delta.s[widgetID].parent !== undefined && delta.s[widgetID].id === undefined) {
```

Comment:
I think I understand why this fixes the stated problem, but I'm not sure I understand what implications this change could have on other situations. I need to better understand how this code relates to the code creating new widgets.

Reply:
In general `applyDeltaToDOM` is called with a first delta that contains all the defaults, then the initial delta as in the JSON and then it only gets _actual_ deltas (so it shouldn't receive a value that's already set). So normally only one delta ever includes `id` - the first one. So flackr used that to detect if it's the first delta so that there is no animation from 0,0 to the actual position.

The code in `receiveDelta` moves all changed widgets to the top surface in case A->B turns into B->A which would otherwise temporarily turn into a cyclic dependency/parentage. In the same block flackr saved the `transform` from the room origin so that it will be applied when adding the widget back to the new parent so it doesn't animate from the room origin (I think)

Does that help?